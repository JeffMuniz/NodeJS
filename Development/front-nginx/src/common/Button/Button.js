import React, { Fragment } from "react";
import { string, func, bool, number } from "prop-types";

import { buttonTypes } from "@enums";
import { SvgIcon } from "@common";
import { If } from "@utils";

import { trackEvent } from "src/modules/Tracking";
import Tooltip from "src/modules/Tooltip";

import { Touchable, IconWrapper } from "./Button.styles";

const Button = ({
  onPress,
  value,
  disabled,
  buttonType,
  loading,
  imgSrc,
  imgWidth,
  imgColor,
  action,
  popupText,
  id,
  ...props
}) => {
  if (id && id.includes("show_docs")) {
    return (
      <Fragment>
        <span data-tip={popupText}>
          <Touchable
            onClick={trackEvent(action, onPress)}
            disabled={loading || disabled}
            buttonType={buttonType}
            id={id}
            {...props}
          >
            <Tooltip html place="top" effect="solid" />
            <span>{loading ? "Carregando" : value}</span>
            <If test={imgSrc}>
              <IconWrapper size={imgWidth} disabled={disabled}>
                <If test={imgColor}>
                  <SvgIcon name={imgSrc} size={imgWidth} fill={imgColor} />
                </If>
                <If test={!imgColor}>
                  <SvgIcon name={imgSrc} size={imgWidth} />
                </If>
              </IconWrapper>
            </If>
          </Touchable>
        </span>
      </Fragment>
    );
  }
  return (
    <Fragment>
      <span data-tip={popupText}>
        <Touchable
          onClick={trackEvent(action, onPress)}
          disabled={loading || disabled}
          buttonType={buttonType}
          id={id}
          {...props}
        >
          <Tooltip html place="top" effect="solid" />
          <If test={imgSrc}>
            <IconWrapper size={imgWidth} disabled={disabled}>
              <If test={imgColor}>
                <SvgIcon name={imgSrc} size={imgWidth} fill={imgColor} />
              </If>
              <If test={!imgColor}>
                <SvgIcon name={imgSrc} size={imgWidth} />
              </If>
            </IconWrapper>
          </If>
          <span>{loading ? "Carregando" : value}</span>
        </Touchable>
      </span>
    </Fragment>
  );
};

Button.propTypes = {
  value: string,
  buttonType: string,
  disabled: bool,
  loading: bool,
  onPress: func,
  imgSrc: string,
  imgWidth: number,
  imgColor: string,
  action: string,
  popupText: string,
  id: string,
};

Button.defaultProps = {
  value: "",
  buttonType: buttonTypes.primary,
  disabled: false,
  loading: false,
  onPress: () => null,
  imgSrc: undefined,
  imgWidth: 16,
  imgColor: null,
  action: null,
  popupText: null,
  id: null,
};

export default Button;
