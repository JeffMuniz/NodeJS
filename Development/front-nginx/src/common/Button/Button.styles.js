import styled from "styled-components";
import { switchStyle, ifStyle } from "@utils";

import {
  blue,
  red,
  white,
  darkenGrey,
  disabledGrey,
  disabledDarkGrey,
} from "src/styles/colors";
import { buttonTypes } from "@enums";

const switchType = switchStyle("buttonType");
const isDisabled = ifStyle("disabled");

const {
  primary,
  warning,
  light,
  lightSecondary,
  link,
  linkSecondary,
  actionButton,
} = buttonTypes;

export const Image = styled.img`
  margin-right: 10px;
  opacity: ${isDisabled("0.5", "1")};
`;

export const Touchable = styled.button`
  display: flex;
  align-items: center;
  justify-content: center;
  cursor: ${isDisabled("not-allowed", "pointer")};
  border-radius: 8px;
  font-size: ${switchType({
    [primary]: isDisabled("14px", "16px"),
    [warning]: isDisabled("14px", "16px"),
    [light]: isDisabled("14px", "16px"),
    [lightSecondary]: isDisabled("14px", "16px"),
    [link]: isDisabled("14px", "16px"),
    [linkSecondary]: "14px",
    [actionButton]: isDisabled("14px", "16px"),
  })};
  font-weight: ${switchType({
    [primary]: "bold",
    [warning]: "bold",
    [light]: "bold",
    [lightSecondary]: "bold",
    [link]: isDisabled("bold", "500"),
    [linkSecondary]: isDisabled("bold", "600"),
    [actionButton]: "bold",
  })};
  padding: 10px 1em;
  text-transform: ${switchType({
    [primary]: "uppercase",
    [warning]: "uppercase",
    [light]: "uppercase",
    [lightSecondary]: "uppercase",
    [link]: isDisabled("uppercase", "none"),
    [linkSecondary]: isDisabled("uppercase", "none"),
    [actionButton]: "uppercase",
  })};
  width: ${switchType({
    [primary]: "100%",
    [warning]: "100%",
    [light]: "100%",
    [lightSecondary]: "100%",
    [link]: isDisabled("100%", "auto"),
    [linkSecondary]: isDisabled("100%", "auto"),
    [actionButton]: "auto",
  })};
  border: ${isDisabled(
    "transparent",
    switchType({
      [primary]: `2px solid ${red}`,
      [warning]: `2px solid ${red}`,
      [light]: `2px solid transparent`,
      [lightSecondary]: `2px solid ${darkenGrey}`,
      [link]: "0px solid transparent",
      [linkSecondary]: "0px solid transparent",
      [actionButton]: "0px solid transparent",
    }),
  )};
  color: ${isDisabled(
    disabledDarkGrey,
    switchType({
      [primary]: `${white}`,
      [warning]: `${white}`,
      [light]: isDisabled(disabledGrey, red),
      [lightSecondary]: `${darkenGrey}`,
      [link]: `${blue}`,
      [linkSecondary]: `${blue}`,
      [actionButton]: `${red}`,
    }),
  )};
  background-color: ${switchType({
    [primary]: isDisabled(disabledGrey, red),
    [warning]: `${red}`,
    [light]: `${white}`,
    [lightSecondary]: `${white}`,
    [link]: "transparent",
    [linkSecondary]: "transparent",
    [actionButton]: "transparent",
  })};
  outline: none;
  line-height: ${switchType({
    [primary]: "normal",
    [warning]: "normal",
    [light]: "normal",
    [lightSecondary]: "normal",
    [link]: "22px",
    [linkSecondary]: "24px",
    [actionButton]: "normal",
  })};
  text-decoration: ${switchType({
    [primary]: "none",
    [warning]: "none",
    [light]: "none",
    [lightSecondary]: "none",
    [link]: "none",
    [linkSecondary]: "underline",
    [actionButton]: "none",
  })};
`;

export const IconWrapper = styled.div`
  width: ${props => props.size}px;
  height: ${props => props.size}px;
  margin: 0 7px;
  justify-content: center;
  align-items: center;
  opacity: ${isDisabled("0.5", "1")};
`;
