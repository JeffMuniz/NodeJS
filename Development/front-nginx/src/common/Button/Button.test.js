import React from "react";
import ShallowRenderer from "react-test-renderer/shallow";
import { buttonTypes } from "@enums";
import TestRenderer from "react-test-renderer";
import Button from "./Button";

jest.mock("@common", () => ({
  SvgIcon: "SvgIcon",
}));

let defaultProps = {};
let testRenderer = null;
let instance = null;

beforeEach(() => {
  defaultProps = {
    value: "btn",
    buttonType: buttonTypes.primary,
    disabled: false,
    loading: false,
    onPress: jest.fn(),
    imgSrc: "teste",
    id: "teste",
  };

  testRenderer = TestRenderer.create(<Button {...defaultProps} />);
  instance = testRenderer.root;
});

describe("Button", () => {
  it("Should render Button", () => {
    // given
    const renderer = new ShallowRenderer();

    // when
    renderer.render(<Button {...defaultProps} />);
    const result = renderer.getRenderOutput();

    // then
    expect(result).toMatchSnapshot();
  });

  it("Should return null when call onPress of default props", () => {
    // given
    const props = {
      value: "btn",
      buttonType: buttonTypes.primary,
      disabled: false,
      loading: false,
      id: "teste",
    };

    // when
    testRenderer.update(<Button {...props} />);

    // then
    expect(instance.props.onPress()).tomacull();
  });

  it("Should change value button when loading is true", () => {
    // given
    const renderer = new ShallowRenderer();

    // when
    renderer.render(<Button {...defaultProps} loading />);
    const result = renderer.getRenderOutput();

    // then
    expect(result).toMatchSnapshot();
  });
});
