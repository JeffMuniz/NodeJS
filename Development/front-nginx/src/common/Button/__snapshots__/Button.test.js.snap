// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Should change value button when loading is true 1`] = `
<UNDEFINED>
  <span
    data-tip={null}
  >
    <styled.button
      buttonType="primary"
      disabled={true}
      id="teste"
      onClick={[Function]}
    >
      <ReactTooltip
        clickable={false}
        effect="solid"
        html={true}
        insecure={true}
        place="top"
        resizeHide={true}
        wrapper="div"
      />
      <If
        test="teste"
      >
        <styled.div
          disabled={false}
          size={16}
        >
          <If
            test={null}
          >
            <SvgIcon
              fill={null}
              name="teste"
              size={16}
            />
          </If>
          <If
            test={true}
          >
            <SvgIcon
              name="teste"
              size={16}
            />
          </If>
        </styled.div>
      </If>
      <span>
        Carregando
      </span>
    </styled.button>
  </span>
</UNDEFINED>
`;

exports[`Button Should render Button 1`] = `
<UNDEFINED>
  <span
    data-tip={null}
  >
    <styled.button
      buttonType="primary"
      disabled={false}
      id="teste"
      onClick={[Function]}
    >
      <ReactTooltip
        clickable={false}
        effect="solid"
        html={true}
        insecure={true}
        place="top"
        resizeHide={true}
        wrapper="div"
      />
      <If
        test="teste"
      >
        <styled.div
          disabled={false}
          size={16}
        >
          <If
            test={null}
          >
            <SvgIcon
              fill={null}
              name="teste"
              size={16}
            />
          </If>
          <If
            test={true}
          >
            <SvgIcon
              name="teste"
              size={16}
            />
          </If>
        </styled.div>
      </If>
      <span>
        btn
      </span>
    </styled.button>
  </span>
</UNDEFINED>
`;
